### PR TITLE
[Diagnostics] Improve diagnostics for invalid type access on existential types.

### PR DESCRIPTION
**Pull request by hborla**
This change improves the error message for invalid associated type and type alias access on a protocol or existential type, and fixes a case of missing diagnostics for invalid type access on a generic parameter that is same-type-constrained to an existential type.

Resolves: rdar://88092865

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckType.cpp [hunk 1]
```diff
@@ -188,6 +187,25 @@ Type TypeResolution::resolveDependentMemberType(
 
   auto *concrete = ref->getBoundDecl();
 
+  if (auto concreteBase = genericSig->getConcreteType(baseTy)) {
+    bool hasUnboundOpener = !!getUnboundTypeOpener();
```
Comment (theblixguy) [inline lib/Sema/TypeCheckType.cpp:191]: Should this be `!`?

Comment (hborla) [reply lib/Sema/TypeCheckType.cpp:191]: Nope, I want the value of `getUnboundTypeOpener` as a `bool`. This `!!` pattern is used elsewhere in TypeCheckType.cpp on `getUnboundTypeOpener` to do the same thing.

Comment (theblixguy) [reply lib/Sema/TypeCheckType.cpp:191]: Oh interesting! I didn’t know we had a `!!`.

Comment (hborla) [reply lib/Sema/TypeCheckType.cpp:191]: I didn't either until I looked at the other call-sites of `TypeChecker::isUnsupportedMemberTypeAccess` 😄

**Review by slavapestov — APPROVED** [review]

**Comment by hborla** [comment]
@swift-ci please smoke test

**Pull request merged**